### ISSUE TITLE
PyFunceble 4.0, the next and future of PyFunceble

### ISSUE THREAD

**Issue by funilrys** [funilrys/PyFunceble#134]
Hello, World!

As some of you may have noticed, a few weeks I slowed down all my activities around here. I got some emails from people asking if I'm okay and if I'm going to continue the development of PyFunceble or other things I wrote, created, or maintain. 

I would like to thank those the people who asked! Yes, I'm good. But for my personal well being, I had to take some days off of any sort of public contribution. It was something I should have done earlier but the machinery (like air-condition or airflow control system) created to fight against the current crisis accelerated the need as they had a bad effect on my well-being.
It's still hard right now but I try to live with it. 

Now to PyFunceble itself, I took some of those days lately, to rethink about it. Indeed, PyFunceble right now is becoming bigger and bigger, but it's still like a "giant script wrapped into several files". Meaning that although it's a good idea the lack of structure because of the history of PyFunceble makes it hard to maintain. 

Therefore, I'm on the road to restructuring and repairing some of the mistakes of the past. In the last few weekends, I started with the idea that PyFunceble should not be a "giant script" and should be maintainable - which is not really the case right now.

------

Here is what I think is important right now.

1. Split CLI/script logic from the actual logic (availability, reputation, syntax checker).
   
   This is my main goal. This means that the "actual logic" - aka what makes PyFunceble (not talking about the file generation part which is part of the CLI/Script) - is independent and does not heavily depend on the configuration. 

   In fact, the CLI/script logic should be the part independent from the configuration file.

2. Provides a more precise and probably better high-level python API.
  
   Right now, the high-level API is incomplete. For example, it's not clear or even possible to do a simple availability, reputation, or syntax check. Just because I made it heavily dependent on the configuration file. My goal is to provide something useful and usable to others without any huge headache like it is right now.

3. Let's get away from the JSON database format. Let's say hello to the CSV or RAW file format.

   Right now - and @dnmTX, @dead-hosts, and @Ultimate-Hosts-Blacklist noticed it - the JSON format is killing PyFunceble because we need to completely load it in memory. And as some of us uses PyFunceble on a huge dataset, the JSON format is just a killer. The CSV or RAW file format will let us read line by line without having to save a whole file in memory. 

4. Let's add support of DNS-over-TLS and DNS-over-HTTPS.

   Right now, we only support DNS-over-TCP and DNS-over-UDP. We should also support the whole family 😄 

5. Let's reimagine the database structure. 

   In the coming weeks (when the rest is implemented), we (@spirillen and I) will try to imagine the future of the MySQL/MariaDB structure. Indeed, the one right now is great but may become inefficient for a distributed environment.

----------

If you think something should be added to this let me know through a friendly comment 😉 
By the way, thanks to @spirillen for answering when I was away! I hope that we will continue to work together for the future of this 😉 

Stay safe and healthy!

**Comment by funilrys** [comment]
Hello, World!
CC: @spirillen, @ZeroDot1, @mitchellkrogza, @DandelionSprout (Please cc all people I forgot)

I made a big part of my work with 4.0.0 available into the [4.0.0-dev](https://github.com/funilrys/PyFunceble/tree/4.0.0-dev) branch. You are highly invited to test it and give us some feedback through a [discussion](https://github.com/funilrys/PyFunceble/discussions/new) or an [issue](https://github.com/funilrys/PyFunceble/issues/new/choose)!

**WARNING** ⚠️ : 
  - Be sure to backup your tables if you are using the MariaDB or the MySQL databases engine.
  - Be sure to backup your `{inactive_db,whois_db}.json` file before starting to play with the alpha version.

**Notice** ℹ️ :
  - The more feedback and discussion we will have, the better it will be once out of alpha!

**Installation Notice** ℹ️:
  - You are invited to install and test it through the following pip installation procedure.
      ```
      $ pip3 install --user git+https://github.com/funilrys/PyFunceble.git@4.0.0-dev#egg=PyFunceble
      ``` 

-----------------------
Roadmap

Conditions to get out of Alpha (Meaning release as `PyFunceble-dev`:
  - [ ] Python tests completed. 
  - [ ] Automation of tests and deployment through GitHub action.
  - [ ] Continuous Integration under Travis CI tested.
  - [ ] Continuous Integration under Gitlab CI tested.

Condition to get out of Beta (Meaning release as `PyFunceble`):
  - [ ] Decisions explained when feedback were given.

**Comment by funilrys** [comment]
Note for those who are wondering: The documentation (until we get out of alpha) can be found [here](https://pyfunceble.readthedocs.io/en/4.0.0-dev/).

Please take into consideration that I'm still trying to update the documentation as I write this.

**Comment by DandelionSprout** [comment]
Minor bug I came across that I presume doesn't warrant its own issue report:

Installing 4.0.0 through the procedure further above in this thread, fails to upgrade the `domain2idna-1.9.1-py3.7.egg` package to `domain2idna-1.12.0`, leading to the error `Fatal Error: module 'domain2idna' has no attribute 'domain2idna'` when running any file. pip3 and pip also claimed that `domain2idna-1.9.1-py3.7.egg` was up to date, which was incorrect according to a glance at PiPy.

The fix was to delete all files named `domain2idna` that also had the terms `37` or `3.7` in them. 
![image](https://user-images.githubusercontent.com/22780683/103142288-fd4a2900-4700-11eb-8190-d3458bac72af.png)

**Comment by funilrys** [comment]
Hello, World!
Hello @mitchellkrogza @DandelionSprout @ZeroDot1 @spirillen,

Just for information, we are now out of alpha. Please test it and let me know if something is wrong.

- Documentation (still ongoing) is now located at http://pyfunceble.readthedocs.io/en/dev.
 
- You can now download PyFunceble 4.0.0 (beta) through: 

    ```
    $ pip3 install --user --pre PyFunceble-dev  ## Without --pre, PIP won't install the beta version. 
    ```

- The `4.0.0-dev` branch is now **stale** and will be removed next week. All new changes will be pushed to the `dev` branch and all new versions will be automatically pushed to the PyPI through our GitHub Actions. **Please note that we don't deliver any conda packages anymore.**


Stay safe and healthy,
Nissar

**Comment by ZeroDot1** [comment]
My test was unfortunately ended very quickly see:  https://github.com/funilrys/PyFunceble/issues/147#issuecomment-756494340

**Comment by ZeroDot1** [comment]
Hi @funilrys Please don't forget to update the checksums in the AUR.
![Bildschirmfoto vom 2021-01-09 12-21-01](https://user-images.githubusercontent.com/28985171/104090197-304dfa00-5275-11eb-83bb-b906bc1403ab.png)

**Comment by ZeroDot1** [comment]
OK, I have now run this for over 1 hour, there is no change, 1 CPU core is permanently 100% used, and there are no files in the output folder.

![Bildschirmfoto vom 2021-01-09 13-15-21](https://user-images.githubusercontent.com/28985171/104092416-1a940100-5284-11eb-9157-ab49c7d5c254.png)

**Comment by funilrys** [comment]
> Hi @funilrys Please don't forget to update the checksums in the AUR.
> ![Bildschirmfoto vom 2021-01-09 12-21-01](https://user-images.githubusercontent.com/28985171/104090197-304dfa00-5275-11eb-83bb-b906bc1403ab.png)

```
==> Erstelle Paket: pyfunceble-dev 4.0.0b4-2 (Sa 09 Jan 2021 14:18:02 CET)
==> Prüfe Laufzeit-Abhängigkeiten...
==> Prüfe Buildtime-Abhängigkeiten...
==> Empfange Quellen...
  -> PyFunceble-dev-4.0.0b4.tar.gz gefunden
  -> LICENSE gefunden
==> Überprüfe source Dateien mit sha256sums...
    PyFunceble-dev-4.0.0b4.tar.gz ... Durchgelaufen
    LICENSE ... Durchgelaufen
==> Entpacke Quellen...
  -> Entpacke PyFunceble-dev-4.0.0b4.tar.gz mit bsdtar
==> Beginne build()...
[...]
```

It's working here 🤔

**Comment by funilrys** [comment]
> OK, I have now run this for over 1 hour, there is no change, 1 CPU core is permanently 100% used, and there are no files in the output folder.
> 
> `pyfunceble -w 4 -a --syntax -f All.txt`
> 
> ![Bildschirmfoto vom 2021-01-09 13-15-21](https://user-images.githubusercontent.com/28985171/104092416-1a940100-5284-11eb-9157-ab49c7d5c254.png)

That's normal. It is still converting the format from json2csv. That job takes a bit of time but once done you will never meet that again.

About the single CPU in json2csv conversion, I have to convert sequentially (1CPU) because instead of loading the JSON file in memory and risk an out of memory error, I read all line of each file and then convert them to our new dataset layout.

**Comment by ZeroDot1** [comment]
The problem with the AUR package is solved, thank you.

**Comment by funilrys** [comment]
If you really don't want it, simply remove those file from your configuration directory: 

- `inactive_db.json` 
- `whois_db.json` (Please don't do this)

**Comment by ZeroDot1** [comment]
OK, that would have taken forever, my inactive_db.json had a size of 525MB.
I just deleted the inactive_db.json and restarted the whole process, now it runs.

**Comment by spirillen** [comment]
Well it took roughly 10 hours to convert the DB with about 265.000 records :clock10: 

For the file conversion, the slowest part is probably the disk...

**Comment by funilrys** [comment]
@spirillen, have you documented https://github.com/funilrys/PyFunceble/issues/134#issuecomment-757191648 ? 🤔

**Comment by spirillen** [comment]
> @spirillen, have you documented [#134 (comment)](https://github.com/funilrys/PyFunceble/issues/134#issuecomment-757191648) ? thinking

Nope, not yet.

Will throw it in the todo list.... <https://github.com/spirillen/PyFunceble/issues/30>

**Comment by ZeroDot1** [comment]
OK, after a few hours an error occurred and the testing process was stopped.

`Fatal Error: fromisoformat: argument must be str`

**Comment by ZeroDot1** [comment]
PyFunceble is running since a while (5,5H) now, it is very slow. To use plain as default see: https://github.com/funilrys/PyFunceble/issues/178 Makes sense when testing very large lists.

**Comment by ZeroDot1** [comment]
Hi @funilrys OK, now it's been running for over 16 hours (including the 4.5 hours from yesterday).
This makes it very clear for me that PyFunceble is not usable for productive use.
PyFunceble needs to become significantly faster, a simple syntax check should not and cannot take so much time.

**Comment by ZeroDot1** [comment]
`--auto-continue` needs a lot of time until the syntax check is continued.

Default setting in the configuration file not changed.

**Comment by spirillen** [comment]
> `--auto-continue` needs a lot of time until the syntax check is continued.
> 
> Default setting in the configuration file not changed.

So by setting `autocontinue: False` you get some descent speed on `--syntax`?

**Comment by ZeroDot1** [comment]
@spirillen @funilrys When I set --auto-continue to false it runs like a rocket with the speed of light.

**Comment by spirillen** [comment]
Can confirm that.... it's tremendous speed enhangement actually, even against the DB

**Comment by ZeroDot1** [comment]
Addition, as soon as the entire list is processed it hangs again.

**Comment by ZeroDot1** [comment]
The syntax check needs only a few seconds if --auto-continue is set to false.
OK, Confirmed, after a second and third test the whole program hangs no completion message.
I think something is not working as it should.

![Uploading Bildschirmfoto vom 2021-01-11 20-46-24.png…]()

**Comment by ZeroDot1** [comment]
I may have found another bug.
When I run `pyfunceble -w 6 --plain -a --syntax -f "All.txt"` invalid URLs are not written to an extra file.

**Comment by ZeroDot1** [comment]
Addition, if I use --plain HOSTS should not be created.

**Comment by spirillen** [comment]
Hey @ZeroDot1 this is because `you disabled the plain files output and only kept the hosts file. I went into the same pitfall my self. the thing is the generation of the DOMAIN (plain) [default is true](https://pyfunceble.readthedocs.io/en/dev/usage/index.html#plain) and I haven't noticed this.

https://github.com/funilrys/PyFunceble/blob/773e5f96746ddf779da2ee68c5b81f0fe212c08c/PyFunceble/data/infrastructure/.PyFunceble_production.yaml#L169-L173

Try this one in stead:

```console
pyfunceble -w 6 -h -a --syntax -f "All.txt"
```

You should then get the plain (domains/status) output.

**Comment by ZeroDot1** [comment]
@funilrys I'm doing a test run after deleting and recreating the entire configuration (except for Whois.CSV), so far it runs smoothly.
If errors occur, I'll get back to you.

**Comment by ZeroDot1** [comment]
Hi @funilrys, 

Everything worked very well for me today and I released the [CoinBlockerLists](https://zerodot1.gitlab.io/CoinBlockerListsWeb/downloads.html) update.
See: https://gitlab.com/ZeroDot1/CoinBlockerLists/-/commit/58bd86449062ccb05fa1fe5bf821623a38fad73b?expanded=1 
Thank you for your very good work it helps me a lot to move the CoinBlockerLists forward.
![Bildschirmfoto vom 2021-01-16 22-47-50](https://user-images.githubusercontent.com/28985171/104824175-f4f09400-584f-11eb-9d23-64d6ea50e762.png)
If I can I will contribute to your project whenever possible.

**Comment by ZeroDot1** [comment]
@funilrys

Possibly a #bug,

After a complete list has been checked and all files have been created, many entries in the output files are duplicated or even more often.

I noticed the problem because the output files were significantly larger than the input files.

**Comment by ZeroDot1** [comment]
Hi @funilrys A new bug, some invalid (incorrect TLD) URLs were not detected.

For examples see the following link: https://gitlab.com/ZeroDot1/CoinBlockerLists/-/commit/6a934f6ab4ba188c3f28c575c9c4d5304fc6c7f7

**Comment by funilrys** [comment]
Hello everyone,

Just a warning from me, in the coming days/weeks the `b` in `4.0.0bXX` will be removed. Meaning that if you have some automatisms that are downloading the latest `pyfunceble-dev` version, you will get `4.0.0`. 
**This is my last warning.** Be sure that you pined your version or you tested the `4.0.0bXX` against your automatisms. A lot has been changed in this major version. 

## How to pin the version

Simply install/upgrade `pyfunceble-dev==X.Y.Z` instead `pyfunceble-dev` (without version).
For example:

- To install the latest available version: `pip install pyfunceble-dev`
- To always install the version `3.3.9`: `pip install pyfunceble-dev==3.3.9`

Feel free to submit any issues!

P.S.: Please mention or warn the people that you know are using this project actively.

cc: @spirillen,  @DandelionSprout, @ZeroDot1, @mitchellkrogza

**Comment by dnmTX** [comment]
How about that none of them lists is updating for months and months. 
I started moving my lists upstream.
Honestly,loosing hope here that this ever will get fixed.
What's the point of all those Ultimate.Hosts....repos if nothing gets updated.
Looking at you @funilrys 👀

**Comment by funilrys** [comment]
@dnmTX It's next on my list ... I just noticed that the central repository was still behind Travis CI (I don't know how it is possible but it was probably still tight to Mitch's personal account). 

I'm coming back but slowly ... Thanks for the @\ping.

**Comment by spirillen** [comment]
> How about that none of them lists is updating for months and months.
> I started moving my lists upstream.
> Honestly,loosing hope here that this ever will get fixed.
> What's the point of all those Ultimate.Hosts....repos if nothing gets updated.
> Looking at you @funilrys eyes

Hey @dnmTX are you sure you posted this the right place?? seem more like it should have been posted in the IHB group with a ref.. (Just saying)

**Comment by DandelionSprout** [comment]
I think that's *far* more of a matter to report at https://github.com/Ultimate-Hosts-Blacklist/Ultimate.Hosts.Blacklist/issues instead.

**Comment by dnmTX** [comment]
i'm posting whever i see fit....how is that for an answer @DandelionSprout  &&  @spirillen  ?????????

**Comment by funilrys** [comment]
`4.0.0` released. Closing.
